util/math_reference.cpp

```diff
@@ -346,7 +356,17 @@ T mul_hi_signed(T a, T b) {
   lo >>= shft;
   hi += lo + (cross1 >> shft) + (cross2 >> shft);
 
-  return (a >> msb) ^ (b >> msb) ? static_cast<T>(~hi) : hi;
+  T result = hi;
+  // if result is negative
+  if ((a >> msb) ^ (b >> msb)) {
+    result = static_cast<T>(~hi);
+    // check that the low half is zero to see if we need to carry
+    if (0 == static_cast<T>(a * b)) {
+      result += 1;
```

Comment:
I need more time to understand this...

Reply:
OK, if the result of the multiplication of 2 non null operands is 0 there is a carry to propagate. If only one operand is 0, result is -1 and then adding 1 set it back to 0.
This looks like it works. :-)